presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdownLogicalOptimizer.java

```diff
@@ -0,0 +1,527 @@
+                                tableName,
+                                hivePartitionResult.getPartitionColumns(),
+                                // remove comments to optimize serialization costs
+                                pruneColumnComments(hivePartitionResult.getDataColumns()),
+                                hivePartitionResult.getTableParameters(),
+                                hivePartitionResult.getPartitions(),
+                                domainPredicate,
+                                decomposedFilter.getRemainingExpression(),
+                                predicateColumns,
+                                hivePartitionResult.getEnforcedConstraint(),
+                                hivePartitionResult.getBucketHandle(),
+                                hivePartitionResult.getBucketFilter(),
+                                true,
+                                createTableLayoutString(session, tableName, hivePartitionResult.getBucketHandle(), hivePartitionResult.getBucketFilter(), decomposedFilter.getRemainingExpression(), domainPredicate))),
+                TRUE_CONSTANT);
+    }
+
+    private abstract class Visitor
+            extends PlanVisitor<PlanNode, Void>
+    {
+        protected final ConnectorSession session;
+        protected final PlanNodeIdAllocator idAllocator;
+        protected final HiveTransactionManager transactionManager;
+
+        Visitor(
+                ConnectorSession session,
+                PlanNodeIdAllocator idAllocator,
+                HiveTransactionManager transactionManager)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        }
+
+        @Override
+        public PlanNode visitPlan(PlanNode node, Void context)
+        {
+            ImmutableList.Builder<PlanNode> children = ImmutableList.builder();
+            boolean changed = false;
+            for (PlanNode child : node.getSources()) {
+                PlanNode newChild = child.accept(this, null);
+                if (newChild != child) {
+                    changed = true;
+                }
+                children.add(newChild);
+            }
+
+            if (!changed) {
+                return node;
+            }
+            return node.replaceChildren(children.build());
+        }
+    }
+
+    private class FilterVisitor
+            extends Visitor
+    {
+        FilterVisitor(ConnectorSession session, PlanNodeIdAllocator idAllocator, HiveTransactionManager transactionManager)
+        {
+            super(session, idAllocator, transactionManager);
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode filter, Void context)
+        {
+            if (!(filter.getSource() instanceof TableScanNode)) {
+                return filter;
```

Comment:
This doesn't seem right. It should be `return new FilterNode(filter.getId(), filter.getSource().accept(this, null), filter.getPredicate());`. Otherwise, the visitor will stop exploring for such case.

Reply:
Nice catch :)